[bug]: Cluster-size estimator omits the item returned by each limit-one LIST

# Description

`K8sResourceHandler.EstimateClusterSize` estimates the number of resources represented by 16 namespaced GVRs by issuing a `LIST` with `limit=1` for each one.

For every successful response it adds only `metadata.remainingItemCount`. Kubernetes defines that field as the number of subsequent items that were not included in the current page, so it does not include the item already returned in `Items`. The estimator consequently omits the returned page from every successful, non-empty LIST.

This report concerns the arithmetic within the existing representative estimate. It does not claim that the 16-GVR estimate covers every resource type in a cluster.

# Environment

OS: `Ubuntu 24.04.3 LTS`  
Version: `master` at [`10e16a2a4e3011d6b25ca06bbe3caee7a7ee71ac`](https://github.com/kubescape/kubescape/commit/10e16a2a4e3011d6b25ca06bbe3caee7a7ee71ac)  
Go: `go1.26.1 linux/amd64`

# Steps To Reproduce

Add this test to `core/pkg/resourcehandler/k8sresources_estimate_test.go`, which already contains `gvrAwareDynamicClient`:

```go
func TestEstimateClusterSize_CountsReturnedItems(t *testing.T) {
	mockClient := &gvrAwareDynamicClient{
		listFunc: func(gvr schema.GroupVersionResource, ctx context.Context, opts metav1.ListOptions) (*unstructured.UnstructuredList, error) {
			assert.Equal(t, int64(1), opts.Limit)
			return &unstructured.UnstructuredList{
				Items: []unstructured.Unstructured{{Object: map[string]any{
					"apiVersion": "v1",
					"kind":       "Example",
					"metadata":   map[string]any{"name": gvr.Resource + "-1"},
				}}},
			}, nil
		},
	}

	handler := &K8sResourceHandler{
		k8s: &k8sinterface.KubernetesApi{DynamicClient: mockClient},
	}

	size, err := handler.EstimateClusterSize(context.Background(), &cautils.ScanInfo{})
	require.NoError(t, err)
	assert.Equal(t, len(namespacedResourcesToEstimate), size,
		"a returned page item must be counted even when remainingItemCount is absent")
}
```

Run:

```bash
go test ./core/pkg/resourcehandler \
  -run '^TestEstimateClusterSize_CountsReturnedItems$' \
  -count=3
```

The failure is deterministic on all three runs:

```text
=== RUN   TestEstimateClusterSize_CountsReturnedItems
    Error:      Not equal:
                expected: 16
                actual  : 0
--- FAIL: TestEstimateClusterSize_CountsReturnedItems (0.00s)
=== RUN   TestEstimateClusterSize_CountsReturnedItems
    Error:      Not equal:
                expected: 16
                actual  : 0
--- FAIL: TestEstimateClusterSize_CountsReturnedItems (0.00s)
=== RUN   TestEstimateClusterSize_CountsReturnedItems
    Error:      Not equal:
                expected: 16
                actual  : 0
--- FAIL: TestEstimateClusterSize_CountsReturnedItems (0.00s)
FAIL
```

All 16 LIST calls succeeded on each run and each returned one item, but the final-page responses had no `remainingItemCount`.

# Expected behavior

For each successful LIST response, the estimator should add both the number of items in the returned page and `remainingItemCount`. The reproduction should return 16.

# Actual Behavior

The estimator adds only `remainingItemCount`. The reproduction returns 0 even though the 16 successful responses contain 16 objects in total.

# Root cause

The loop increments the number of successful requests but accounts only for items after the current page:

```go
ok++
if rc := result.GetRemainingItemCount(); rc != nil {
	total += int(*rc)
}
```

For a paginated LIST response, the represented collection size is:

```text
len(result.Items) + remainingItemCount
```

When `remainingItemCount` is absent, the returned `Items` are still known and should still be counted.

# Impact

With the current `limit=1` requests, the estimate is low by one for each successfully listed GVR that returns an item. The current representative set contains 16 GVRs, so the maximum arithmetic undercount from this bug is 16.

Kubescape uses this estimate to auto-enable streaming when the configured large-cluster threshold is exceeded (2500 by default). An estimate close enough to that threshold can therefore remain on the non-streaming path when the correctly calculated representative total would exceed it. This report does not claim a specific memory failure or an unbounded undercount.

# Suggested fix

Add `len(result.Items)` for every successful LIST, then add `remainingItemCount` when it is present:

```go
ok++
total += len(result.Items)
if rc := result.GetRemainingItemCount(); rc != nil {
	total += int(*rc)
}
```

Update the existing estimate fixtures so a response with a positive `remainingItemCount` also contains the item returned by the `limit=1` page. Add regression coverage for:

- one returned item with no `remainingItemCount`;
- one returned item plus a positive `remainingItemCount`;
- successful empty LISTs, which should still produce a valid zero estimate;
- partial per-GVR LIST errors, preserving the existing error-tolerance behavior.

# Additional context

The Kubernetes API documentation for [retrieving large result sets in chunks](https://kubernetes.io/docs/reference/using-api/api-concepts/#retrieving-large-results-sets-in-chunks) defines `remainingItemCount` as the number of subsequent items not included in the current response.

PR [#2645](https://github.com/kubescape/kubescape/pull/2645) introduced the estimator and the automatic streaming decision. The per-GVR error handling and the all-LISTs-failed error behavior are separate and need not change for this fix.

I could not find an open issue or pull request covering this specific failure mode.

--- a/core/pkg/resourcehandler/k8sresources.go
+++ b/core/pkg/resourcehandler/k8sresources.go
@@ -1001,8 +1001,8 @@ var namespacedResourcesToEstimate = []schema.GroupVersionResource{
 }
 
 // EstimateClusterSize estimates the number of namespaced resources in the
-// cluster by issuing metadata-only LIST requests (limit=1) to the API server
-// and summing the remainingItemCount from each response. A per-GVR failure
+// cluster by issuing LIST requests (limit=1) to the API server and summing the
+// returned items and remainingItemCount from each response. A per-GVR failure
 // (type unavailable, or the service account lacking read access) is tolerated,
 // but if no representative type can be listed at all the estimate is useless —
 // returning (0, nil) would be indistinguishable from a genuinely small cluster
@@ -1021,6 +1021,7 @@ func (k8sHandler *K8sResourceHandler) EstimateClusterSize(ctx context.Context, s
 			continue
 		}
 		ok++
+		total += len(result.Items)
 		if rc := result.GetRemainingItemCount(); rc != nil {
 			total += int(*rc)
 		}

--- a/core/pkg/resourcehandler/k8sresources_estimate_test.go
+++ b/core/pkg/resourcehandler/k8sresources_estimate_test.go
@@ -56,14 +56,59 @@ func TestEstimateClusterSize_NilDynamicClient(t *testing.T) {
 	assert.Equal(t, 0, size)
 }
 
-func newListWithRemaining(n int64) *unstructured.UnstructuredList {
+func newLimitOneList() *unstructured.UnstructuredList {
 	return &unstructured.UnstructuredList{
-		Object: map[string]any{
-			"metadata": map[string]any{
-				"remainingItemCount": n,
-			},
+		Items: []unstructured.Unstructured{{Object: map[string]any{
+			"apiVersion": "v1",
+			"kind":       "Example",
+			"metadata":   map[string]any{"name": "first"},
+		}}},
+	}
+}
+
+func newLimitOneListWithRemaining(n int64) *unstructured.UnstructuredList {
+	list := newLimitOneList()
+	list.Object = map[string]any{
+		"metadata": map[string]any{
+			"remainingItemCount": n,
 		},
 	}
+	return list
+}
+
+func TestEstimateClusterSize_CountsReturnedItems(t *testing.T) {
+	mockClient := &gvrAwareDynamicClient{
+		listFunc: func(_ schema.GroupVersionResource, ctx context.Context, opts metav1.ListOptions) (*unstructured.UnstructuredList, error) {
+			assert.Equal(t, int64(1), opts.Limit)
+			return newLimitOneList(), nil
+		},
+	}
+
+	handler := &K8sResourceHandler{
+		k8s: &k8sinterface.KubernetesApi{DynamicClient: mockClient},
+	}
+
+	size, err := handler.EstimateClusterSize(context.Background(), &cautils.ScanInfo{})
+	require.NoError(t, err)
+	assert.Equal(t, len(namespacedResourcesToEstimate), size,
+		"a returned page item must be counted even when remainingItemCount is absent")
+}
+
+func TestEstimateClusterSize_SuccessfulEmptyLists(t *testing.T) {
+	mockClient := &gvrAwareDynamicClient{
+		listFunc: func(_ schema.GroupVersionResource, ctx context.Context, opts metav1.ListOptions) (*unstructured.UnstructuredList, error) {
+			assert.Equal(t, int64(1), opts.Limit)
+			return &unstructured.UnstructuredList{}, nil
+		},
+	}
+
+	handler := &K8sResourceHandler{
+		k8s: &k8sinterface.KubernetesApi{DynamicClient: mockClient},
+	}
+
+	size, err := handler.EstimateClusterSize(context.Background(), &cautils.ScanInfo{})
+	require.NoError(t, err)
+	assert.Zero(t, size, "successfully listed empty resource types form a valid zero-size estimate")
 }
 
 func TestEstimateClusterSize_SmallCluster(t *testing.T) {
@@ -81,7 +126,7 @@ func TestEstimateClusterSize_SmallCluster(t *testing.T) {
 			case "configmaps":
 				n = 20
 			}
-			return newListWithRemaining(n), nil
+			return newLimitOneListWithRemaining(n), nil
 		},
 	}
 
@@ -91,14 +136,14 @@ func TestEstimateClusterSize_SmallCluster(t *testing.T) {
 
 	size, err := handler.EstimateClusterSize(context.Background(), &cautils.ScanInfo{})
 	require.NoError(t, err)
-	// 50+10+5+20 + (12 other GVRs * 10 each) = 85 + 120 = 205
-	assert.Equal(t, 205, size)
+	// 50+10+5+20 + (12 other GVRs * 10 each) + 16 returned items = 221
+	assert.Equal(t, 221, size)
 }
 
 func TestEstimateClusterSize_LargeCluster(t *testing.T) {
 	mockClient := &gvrAwareDynamicClient{
 		listFunc: func(gvr schema.GroupVersionResource, ctx context.Context, opts metav1.ListOptions) (*unstructured.UnstructuredList, error) {
-			return newListWithRemaining(500), nil
+			return newLimitOneListWithRemaining(500), nil
 		},
 	}
 
@@ -108,8 +153,8 @@ func TestEstimateClusterSize_LargeCluster(t *testing.T) {
 
 	size, err := handler.EstimateClusterSize(context.Background(), &cautils.ScanInfo{})
 	require.NoError(t, err)
-	// 16 GVRs * 500 each = 8000
-	assert.Equal(t, 8000, size)
+	// 16 GVRs * (500 remaining + 1 returned) = 8016
+	assert.Equal(t, 8016, size)
 }
 
 func TestEstimateClusterSize_ListErrors(t *testing.T) {
@@ -118,7 +163,7 @@ func TestEstimateClusterSize_ListErrors(t *testing.T) {
 			if gvr.Resource == "pods" {
 				return nil, errors.New("API server error")
 			}
-			return newListWithRemaining(100), nil
+			return newLimitOneListWithRemaining(100), nil
 		},
 	}
 
@@ -128,8 +173,8 @@ func TestEstimateClusterSize_ListErrors(t *testing.T) {
 
 	size, err := handler.EstimateClusterSize(context.Background(), &cautils.ScanInfo{})
 	require.NoError(t, err)
-	// 15 GVRs * 100 = 1500 (pods error is skipped)
-	assert.Equal(t, 1500, size)
+	// 15 GVRs * (100 remaining + 1 returned) = 1515 (pods error is skipped)
+	assert.Equal(t, 1515, size)
 }
 
 func TestEstimateClusterSize_AllListErrors(t *testing.T) {
@@ -176,10 +221,10 @@ func TestEstimateClusterSize_NilRemainingItemCount(t *testing.T) {
 	mockClient := &gvrAwareDynamicClient{
 		listFunc: func(gvr schema.GroupVersionResource, ctx context.Context, opts metav1.ListOptions) (*unstructured.UnstructuredList, error) {
 			if gvr.Resource == "pods" {
-				// No metadata at all — GetRemainingItemCount returns nil
-				return &unstructured.UnstructuredList{}, nil
+				// A complete one-item page has no remainingItemCount.
+				return newLimitOneList(), nil
 			}
-			return newListWithRemaining(200), nil
+			return newLimitOneListWithRemaining(200), nil
 		},
 	}
 
@@ -189,6 +234,6 @@ func TestEstimateClusterSize_NilRemainingItemCount(t *testing.T) {
 
 	size, err := handler.EstimateClusterSize(context.Background(), &cautils.ScanInfo{})
 	require.NoError(t, err)
-	// 15 GVRs * 200 = 3000 (pods with nil remainingItemCount is skipped)
-	assert.Equal(t, 3000, size)
+	// 15 GVRs * (200 remaining + 1 returned) + 1 returned pod = 3016.
+	assert.Equal(t, 3016, size)
 }

